File: python/tests/test_v1_1_encryption_parity.py
"""v1.1 parity: encrypt → close → reopen → is_encrypted → decrypt → read.

Covers the two bugs reported in the MPEG-O-MCP-Server M5 handoff:
 * Issue A: SpectralDataset.is_encrypted / encrypted_algorithm lost state
   across close/reopen.
 * Issue B: decrypt_with_key(key) left spec.intensity_array unusable
   because the in-memory channel cache was never rehydrated.

Byte-level parity with the ObjC/Java equivalents is exercised by
test_cross_compat.py; this file pins the Python surface so we notice
if the accessors regress without needing an ObjC fixture.
"""
from __future__ import annotations

from pathlib import Path

import h5py
import numpy as np

from mpeg_o import SpectralDataset
from mpeg_o.enums import AcquisitionMode, EncryptionLevel


def _write_fixture(path: Path, run_names: list[str]) -> None:
    with h5py.File(path, "w") as f:
        f.attrs["mpeg_o_format_version"] = "1.1"
        study = f.create_group("study")
        runs = study.create_group("ms_runs")
        runs.attrs["_run_names"] = ",".join(run_names)
        for rname in run_names:
            g = runs.create_group(rname)
            g.attrs["acquisition_mode"] = np.int64(AcquisitionMode.MS1_DDA)
            g.attrs["spectrum_class"] = "MPGOMassSpectrum"
            idx = g.create_group("spectrum_index")
            idx.create_dataset("offsets", data=np.array([0], dtype="<u8"))
            idx.create_dataset("lengths", data=np.array([4], dtype="<u4"))
            idx.create_dataset("retention_times", data=np.array([0.0], dtype="<f8"))
            idx.create_dataset("ms_levels", data=np.array([1], dtype="<i4"))
            idx.create_dataset("polarities", data=np.array([1], dtype="<i4"))
            idx.create_dataset("precursor_mzs", data=np.array([0.0], dtype="<f8"))
            idx.create_dataset("precursor_charges", data=np.array([0], dtype="<i4"))
            idx.create_dataset("base_peak_intensities", data=np.array([0.0], dtype="<f8"))
            sc = g.create_group("signal_channels")
            sc.attrs["channel_names"] = "mz,intensity"
            sc.create_dataset("mz_values",
                              data=np.array([100.0, 200.0, 300.0, 400.0], dtype="<f8"))
            sc.create_dataset("intensity_values",
                              data=np.array([1.0, 2.0, 3.0, 4.0], dtype="<f8"))


def test_issue_a_encrypted_state_survives_close_reopen(tmp_path: Path) -> None:
    path = tmp_path / "a.mpgo"
    _write_fixture(path, ["run_0001"])
    key = bytes(range(32))

    with SpectralDataset.open(str(path), writable=True) as ds:
        assert not ds.is_encrypted
        assert ds.encrypted_algorithm == ""
        ds.encrypt_with_key(key, EncryptionLevel.DATASET)
        assert ds.is_encrypted, "must flip in-memory immediately after encrypt"
        assert ds.encrypted_algorithm == "aes-256-gcm"

    # The critical assertion: state must persist to disk and be visible
    # to a fresh reader (this was Issue A).
    with SpectralDataset.open(str(path)) as ds:
        assert ds.is_encrypted
        assert ds.encrypted_algorithm == "aes-256-gcm"


def test_issue_b_decrypt_rehydrates_intensity_for_spectra(tmp_path: Path) -> None:
    path = tmp_path / "b.mpgo"
    _write_fixture(path, ["run_0001"])
    key = bytes(range(32))
    expected = np.array([1.0, 2.0, 3.0, 4.0], dtype="<f8")

    with SpectralDataset.open(str(path), writable=True) as ds:
        ds.encrypt_with_key(key, EncryptionLevel.DATASET)

    with SpectralDataset.open(str(path)) as ds:
        assert ds.is_encrypted
        run = ds.ms_runs["run_0001"]
        result = run.decrypt_with_key(key)
        # decrypt_with_key still returns the raw bytes for backward
        # compatibility with callers that want to do their own parsing.
        raw = np.frombuffer(result, dtype="<f8")
        np.testing.assert_array_equal(raw, expected)

        # Issue B: the spectrum API must see the decrypted intensities
        # without the caller having to parse bytes themselves.
        spec = run[0]
        np.testing.assert_array_equal(spec.intensity_array.data, expected)


def test_issue_b_dataset_level_decrypt_rehydrates_all_runs(tmp_path: Path) -> None:
    path = tmp_path / "c.mpgo"
    _write_fixture(path, ["run_A", "run_B"])
    key = bytes(range(32))
    expected = np.array([1.0, 2.0, 3.0, 4.0], dtype="<f8")

    with SpectralDataset.open(str(path), writable=True) as ds:
        ds.encrypt_with_key(key, EncryptionLevel.DATASET)

    with SpectralDataset.open(str(path)) as ds:
        assert ds.is_encrypted
        ds.decrypt_with_key(key)
        for rname in ("run_A", "run_B"):
            run = ds.ms_runs[rname]
            np.testing.assert_array_equal(run[0].intensity_array.data, expected,
                                          err_msg=f"{rname}: intensity mismatch")

File: objc/Tests/TestV11EncryptionParity.m
// v1.1 parity: encrypt -> close -> reopen -> isEncrypted -> decrypt -> read.
//
// Covers the two bugs reported in the TTI-O-MCP-Server M5 handoff:
//   * Issue A: TTIOSpectralDataset.isEncrypted / .encryptedAlgorithm lost
//     state across close/reopen because the @encrypted root attribute was
//     not being persisted.
//   * Issue B: -decryptWithKey:error: left TTIOMassSpectrum.intensityArray
//     unusable because the in-memory channel cache was never rehydrated
//     with plaintext after a round trip through disk.
//
// Mirrors python/tests/test_v1_1_encryption_parity.py and
// java/src/test/java/.../ProtectionTest.java#v11* so a regression in any
// one implementation is caught by the parity suite before a v1.1 release.
#pragma clang diagnostic ignored "-Wdeprecated-declarations"

#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Core/TTIOSignalArray.h"
#import "Spectra/TTIOMassSpectrum.h"
#import "Run/TTIOAcquisitionRun.h"
#import "Run/TTIOInstrumentConfig.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "ValueClasses/TTIOEncodingSpec.h"
#import "ValueClasses/TTIOEnums.h"
#import <unistd.h>
#import <assert.h>

static NSString *v11path(NSString *suffix)
{
    return [NSString stringWithFormat:@"/tmp/ttio_test_v11_parity_%d_%@.tio",
            (int)getpid(), suffix];
}

static TTIOSignalArray *v11F64(const double *values, NSUInteger n)
{
    NSData *buf = [NSData dataWithBytes:values length:n * sizeof(double)];
    TTIOEncodingSpec *enc =
        [TTIOEncodingSpec specWithPrecision:TTIOPrecisionFloat64
                       compressionAlgorithm:TTIOCompressionZlib
                                  byteOrder:TTIOByteOrderLittleEndian];
    return [[TTIOSignalArray alloc] initWithBuffer:buf
                                            length:n
                                          encoding:enc
                                              axis:nil];
}

static TTIOAcquisitionRun *v11MakeOneSpectrumRun(const double *intensity, NSUInteger n)
{
    double mzBuf[4] = {100.0, 200.0, 300.0, 400.0};
    assert(n == 4 && "fixture assumes 4-point spectrum");
    TTIOMassSpectrum *spec =
        [[TTIOMassSpectrum alloc] initWithMzArray:v11F64(mzBuf, n)
                                   intensityArray:v11F64(intensity, n)
                                          msLevel:1
                                         polarity:TTIOPolarityPositive
                                       scanWindow:nil
                                    indexPosition:0
                                  scanTimeSeconds:0.0
                                      precursorMz:0.0
                                  precursorCharge:0
                                            error:NULL];
    TTIOInstrumentConfig *cfg =
        [[TTIOInstrumentConfig alloc] initWithManufacturer:@""
                                                     model:@""
                                              serialNumber:@""
                                                sourceType:@""
                                              analyzerType:@""
                                              detectorType:@""];
    return [[TTIOAcquisitionRun alloc] initWithSpectra:@[spec]
                                       acquisitionMode:TTIOAcquisitionModeMS1DDA
                                      instrumentConfig:cfg];
}

static NSString *v11WriteFixture(NSString *suffix, const double *intensity)
{
    TTIOSpectralDataset *ds =
        [[TTIOSpectralDataset alloc] initWithTitle:@"v11_parity"
                                isaInvestigationId:@""
                                            msRuns:@{@"run_0001": v11MakeOneSpectrumRun(intensity, 4)}
                                           nmrRuns:@{}
                                   identifications:@[]
                                   quantifications:@[]
                                 provenanceRecords:@[]
                                       transitions:nil];
    NSString *path = v11path(suffix);
    unlink([path fileSystemRepresentation]);
    NSError *err = nil;
    if (![ds writeToFilePath:path error:&err]) {
        NSLog(@"v11 fixture write failed: %@", err);
        return nil;
    }
    [ds closeFile];
    return path;
}

static NSData *v11KnownKey(void)
{
    uint8_t k[32];
    for (int i = 0; i < 32; i++) k[i] = (uint8_t)i;
    return [NSData dataWithBytes:k length:32];
}

// ---- Issue A: is_encrypted / encrypted_algorithm survive close/reopen ----
static void v11TestEncryptedStateSurvivesCloseReopen(void)
{
    double intensity[4] = {1.0, 2.0, 3.0, 4.0};
    NSString *path = v11WriteFixture(@"issueA", intensity);
    PASS(path != nil, "Issue A: fixture writes");

    NSData *key = v11KnownKey();
    NSError *err = nil;

    // Encrypt via a fresh read handle, mirroring the Python/Java parity
    // sequence. closeFile is required so the encryption manager can
    // reopen the file read-write.
    TTIOSpectralDataset *writer =
        [TTIOSpectralDataset readFromFilePath:path error:&err];
    PASS(writer != nil, "Issue A: initial read succeeds");
    PASS(!writer.isEncrypted,
         "Issue A: plaintext dataset reports isEncrypted == NO");
    PASS([writer.encryptedAlgorithm isEqualToString:@""],
         "Issue A: plaintext dataset reports empty encryptedAlgorithm");

    [writer closeFile];
    err = nil;
    BOOL enc = [writer encryptWithKey:key
                                level:TTIOEncryptionLevelDataset
                                error:&err];
    PASS(enc, "Issue A: encryptWithKey:level: succeeds");
    PASS(writer.isEncrypted,
         "Issue A: isEncrypted flips YES in-memory immediately after encrypt");
    PASS([writer.encryptedAlgorithm isEqualToString:@"aes-256-gcm"],
         "Issue A: encryptedAlgorithm is aes-256-gcm after encrypt");
    [writer closeFile];

    // Critical assertion: state must persist to disk and be visible to a
    // fresh reader (this was Issue A).
    err = nil;
    TTIOSpectralDataset *reader =
        [TTIOSpectralDataset readFromFilePath:path error:&err];
    PASS(reader != nil, "Issue A: reopen after encrypt succeeds");
    PASS(reader.isEncrypted,
         "Issue A: reopened dataset reports isEncrypted == YES");
    PASS([reader.encryptedAlgorithm isEqualToString:@"aes-256-gcm"],
         "Issue A: reopened dataset reports aes-256-gcm");
    [reader closeFile];

    unlink([path fileSystemRepresentation]);
}

// ---- Issue B: decrypt rehydrates in-memory intensity channel ----
static void v11TestDecryptRehydratesIntensity(void)
{
    double intensity[4] = {1.0, 2.0, 3.0, 4.0};
    NSString *path = v11WriteFixture(@"issueB", intensity);
    PASS(path != nil, "Issue B: fixture writes");

    NSData *key = v11KnownKey();
    NSError *err = nil;

    TTIOSpectralDataset *writer =
        [TTIOSpectralDataset readFromFilePath:path error:&err];
    PASS(writer != nil, "Issue B: initial read succeeds");
    [writer closeFile];
    err = nil;
    PASS([writer encryptWithKey:key
                          level:TTIOEncryptionLevelDataset
                          error:&err],
         "Issue B: encryptWithKey:level: succeeds");
    [writer closeFile];

    err = nil;
    TTIOSpectralDataset *sealed =
        [TTIOSpectralDataset readFromFilePath:path error:&err];
    PASS(sealed.isEncrypted, "Issue B: reopened dataset is encrypted");
    err = nil;
    PASS([sealed decryptWithKey:key error:&err],
         "Issue B: decryptWithKey: succeeds on reopened dataset");

    TTIOAcquisitionRun *run = sealed.msRuns[@"run_0001"];
    PASS(run != nil, "Issue B: run_0001 present after decrypt");

    err = nil;
    id specAny = [run spectrumAtIndex:0 error:&err];
    PASS([specAny isKindOfClass:[TTIOMassSpectrum class]],
         "Issue B: spectrumAtIndex:0: returns TTIOMassSpectrum");

    TTIOMassSpectrum *spec = (TTIOMassSpectrum *)specAny;
    TTIOSignalArray *intensityArr = spec.intensityArray;
    PASS(intensityArr != nil,
         "Issue B: decrypted spectrum exposes non-nil intensityArray "
         "(this was the KeyError-equivalent before the fix)");
    PASS(intensityArr.length == 4,
         "Issue B: decrypted intensity length matches original");

    if (intensityArr.length == 4 && intensityArr.buffer.length == 4 * sizeof(double)) {
        const double *out = (const double *)intensityArr.buffer.bytes;
        BOOL match = YES;
        for (NSUInteger i = 0; i < 4; i++) {
            if (out[i] != intensity[i]) { match = NO; break; }
        }
        PASS(match, "Issue B: decrypted intensity bytes equal plaintext");
    } else {
        PASS(NO, "Issue B: decrypted intensity has unexpected shape");
    }

    [sealed closeFile];
    unlink([path fileSystemRepresentation]);
}

void testV11EncryptionParity(void)
{
    v11TestEncryptedStateSurvivesCloseReopen();
    v11TestDecryptRehydratesIntensity();
}
